map/extrapolation/extrapolator.cpp

```diff
@@ -0,0 +1,195 @@
+#include "map/extrapolation/extrapolator.hpp"
+
+#include "geometry/distance_on_sphere.hpp"
+
+#include "platform/platform.hpp"
+
+#include "base/logging.hpp"
+
+#include <chrono>
+#include <memory>
+
+namespace
+{
+// If the difference of values between two instances of GpsInfo is greater
+// than the appropriate constant below the extrapolator will be switched off for
+// these two instances of GpsInfo.
+double constexpr kMaxExtrapolationSpeedMPS = 85.0;
+double constexpr kMaxExtrapolationDistMeters = 120.0;
+double constexpr kMaxExtrapolationTimeSeconds = 2.1;
+
+class LinearExtrapolator
+{
+public:
+  LinearExtrapolator(uint64_t timeBetweenMs, uint64_t timeAfterMs)
+    : m_timeBetweenMs(timeBetweenMs), m_timeAfterMs(timeAfterMs)
+  {
+    CHECK_NOT_EQUAL(m_timeBetweenMs, 0, ());
+  }
+
+  double Extrapolate(double x1, double x2)
+  {
+    return x2 + ((x2 - x1) / m_timeBetweenMs) * m_timeAfterMs;
+  }
+
+private:
+  uint64_t m_timeBetweenMs;
+  uint64_t m_timeAfterMs;
+};
+}  // namespace
+
+namespace extrapolation
+{
+using namespace std;
+
+location::GpsInfo LinearExtrapolation(location::GpsInfo const & gpsInfo1,
+                                      location::GpsInfo const & gpsInfo2,
+                                      uint64_t timeAfterPoint2Ms)
+{
+  CHECK_LESS(gpsInfo1.m_timestamp, gpsInfo2.m_timestamp, ());
+  auto const timeBetweenPointsMs =
+      static_cast<uint64_t>((gpsInfo2.m_timestamp - gpsInfo1.m_timestamp) * 1000);
+
+  location::GpsInfo result = gpsInfo2;
+  LinearExtrapolator e(timeBetweenPointsMs, timeAfterPoint2Ms);
+
+  result.m_timestamp += static_cast<double>(timeAfterPoint2Ms) / 1000.0;
+  result.m_longitude = e.Extrapolate(gpsInfo1.m_longitude, gpsInfo2.m_longitude);
```

Comment:
Hmm, probably I wrote about this issue before, but can't find your reply...
Whatever, lat/lon are periodic variables, therefore linear extrapolations don't work for it well, especially in corner cases. As for bearing.

Reply:
Yes, thanks. Lat lon are periodic variable as well. I'll do some with it within this PR.

Reply:
In this PR there's a method for switching off extrapolation if location is not good (rare or next location is too far from previous one). I added switching off extrapolation in current location is too close to meredian 180, north or south pole. Taking into account that very few number of roads in this area and in this edge case the application continue working based on gps coords, I guess it's a possible solution.